### ISSUE TITLE
New Rule: no-async-actions

### ISSUE THREAD

**Issue by Turbo87** [ember-cli/eslint-plugin-ember#418]
see http://ember-concurrency.com/docs/tutorial (scroll down to Version 4)

**Bad:**

```js
actions: {
  async handleClick() {
    // ...
  }
}
```
```js
actions: {
  handleClick() {
    return something.then(() => { /* ... */ });
  }
}
```
```js
@action
async handleClick() {
  // ...
}
```
```js
@action
handleClick() {
  return something.then(() => { /* ... */ });
}
```

**Good:**

```js
actions: {
  handleClick() {
    return nothingOrSomethingThatIsNotAPromise;
  }
}
```

**Comment by rajasegar** [comment]
@Turbo87 For the decorators, should we need to configure some parser options i am getting the following error. Any help would be appreciated.
<img width="605" alt="Screen Shot 2019-05-25 at 5 51 47 PM" src="https://user-images.githubusercontent.com/2083049/58369289-e97f8500-7f15-11e9-93db-21f5f45e3ee1.png">

**Comment by Turbo87** [comment]
good question, I don't know 😅

we might have to configure the babel-eslint parser for those cases

**Comment by rajasegar** [comment]
@Turbo87 I need your help in formulating the docs for this rule
Is it on the same lines with `eslint/no-async-promise-executor` rule in eslint
https://github.com/eslint/eslint/blob/master/docs/rules/no-async-promise-executor.md

@bmish needs some clarifications here 
https://github.com/ember-cli/eslint-plugin-ember/pull/424#discussion_r289675593

**Comment by Turbo87** [comment]
> Is it on the same lines with `eslint/no-async-promise-executor` rule in eslint

no, it's unrelated

> see http://ember-concurrency.com/docs/tutorial (scroll down to Version 4)

⬆️

**Comment by rajasegar** [comment]
@Turbo87 Then, can you please help me in coming up with a proper brief about the rule

**Comment by rajasegar** [comment]
@bmish Regarding the error above, any help would be appreciated.

**Comment by Turbo87** [comment]
as written in the initial post:

> see http://ember-concurrency.com/docs/tutorial (scroll down to Version 4)

tl;dr: using async actions can lead to memory leaks and application errors if you don't check for `isDestroying` and `isDestroyed` after each async step

**Comment by rajasegar** [comment]
Sounds good to a brief, but do we also need to mention how to fix this error or any recommended practice to avoid async actions.

**Comment by mongoose700** [comment]
I don't think this rule makes sense. There are going to be cases where we want to trigger asynchronous behavior through actions, but this doesn't address how we would do so. The `good` example is returning something other than a promise, but if we just had a promise that didn't get returned, this lint rule would just ignore it.

Given the stated goal is to avoid errors from not checking `isDestroying` or `isDestroyed`, it would make more sense for the lint rule to focus on that. If someone did correctly check those, it would still fail the lint rule as it's currently written. And if someone does handle it correctly, they shouldn't have to then manually disable the rule.

**Comment by Turbo87** [comment]
@mongoose700 this rule will be optional, if you don't want it, then you don't have to enable it.

> There are going to be cases where we want to trigger asynchronous behavior through actions

you can trigger it, there is nothing wrong with that. it's just that doing anything with `this` after the async behavior has finished can be very dangerous.

**Comment by mongoose700** [comment]
> @mongoose700 this rule will be optional, if you don't want it, then you don't have to enable it.

Even so, I think it should do a better job at targeting the behavior it's intended to discourage.

> you can trigger it, there is nothing wrong with that. it's just that doing anything with `this` after the async behavior has finished can be very dangerous.

Then the rule should be focused on anything that may modify `this` asynchronously, not on just whether the function is `async` or returns a promise. If someone has an action that violates this rule, they are more likely to fix it by making it stop returning the promise than by adding the proper checks, especially since the error message is simply `Do not use async actions`. Do you believe that async actions should be avoided entirely? The rule would make the entire example in the tutorial invalid.

**Comment by Turbo87** [comment]
> Do you believe that async actions should be avoided entirely? 

yes, unless you know about `isDestroying` and `isDestroyed` and promise to *always* apply them properly.

or alternatively one can use `ember-concurrency` which avoids these issues completely...

> I think it should do a better job at targeting the behavior it's intended to discourage

you are very welcome to work on improving this rule 😉

**Comment by bendemboski** [comment]
Is there a reason this is targeting actions specifically, as opposed to any member functions of `EmberObject` or glimmer Component sub-classes? Seems like the pitfalls aren't unique to "actions," and the `@action` decorator no longer even really implies that it's called from a template, so targeting "actions" seems a little arbitrary?

**Comment by mansona** [comment]
@bendemboski there is a reason why targeting actions is a good idea. The main reason why you would want to add `@action` in the first place is because you want to have the ability to access `this.` in your function. If you access `this.` after an await in an async action then there is a possibility that `this.` doesn't exist any more.

If you use ember-concurrency this is handled for you automatically so I think it's reasonable to just flat out warn against using `async` with actions. There may be an argument to make this rule super smart and check if you're accessing `this.` after an `await` call, but that seems a bit overkill, and if you're wanting async with your actions you will likely have a much better time just using ember-concurrency.
